salt/loader/lazy.py

```diff
@@ -790,8 +790,17 @@ def _load_module(self, name):
                     spec = file_finder.find_spec(mod_namespace)
                     if spec is None:
                         raise ImportError()
-                    mod = importlib.util.module_from_spec(spec)
-                    spec.loader.exec_module(mod)
+                    if mod_namespace in sys.modules:
+                        mod = sys.modules[mod_namespace]
+                        _update_module_attrs(mod, spec)
+                    else:
+                        mod = importlib.util.module_from_spec(spec)
+                    # We need to execute the module in this loader's context in case
+                    # we're reloading an already initialized module with NamedLoaderContext instances.
+                    # Accessing one of the Salt-specific dunders during initialization would throw a NameError usually,
+                    # but during reload without an active loader, NamedLoaderContexts resolve to None, which unexpectedly
+                    # raises a TypeError on access instead of a NameError.
+                    self.run(spec.loader.exec_module, mod)
```

Comment:
For Salt core, the only instance of this issue came up in unit tests for `salt/renderers/pyobjects.py`: https://github.com/saltstack/salt/blob/3ce11a76c76e7cca536e7134188d9a9c04a82b49/salt/renderers/pyobjects.py#L336-L339

It's still a change of behavior from 3006.x to 3007.x, so I opted for `self.run` instead of fixing the singular instance by adding a `TypeError` to the caught exceptions.